plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/worker/KtLintWorkAction.kt

```diff
@@ -0,0 +1,121 @@
+import java.util.ServiceLoader
+
+@Suppress("UnstableApiUsage")
+abstract class KtLintWorkAction : WorkAction<KtLintWorkAction.KtLintWorkParameters> {
+    override fun execute() {
+        val ruleSets = loadRuleSetsAndFilterThem(
+            parameters.enableExperimental.getOrElse(false),
+            parameters.disabledRules.getOrElse(emptySet())
+        )
+
+        val additionalEditorConfig = parameters
+            .additionalEditorconfigFile
+            .orNull
+            ?.asFile
+            ?.absolutePath
+        val userData = generateUserData()
+        val debug = parameters.debug.get()
+        val formatSource = parameters.formatSource.getOrElse(false)
+
+        val result = mutableListOf<LintErrorResult>()
+
+        parameters.filesToLint.files.forEach {
+            val errors = mutableListOf<Pair<SerializableLintError, Boolean>>()
+            val ktLintParameters = KtLint.Params(
+                fileName = it.absolutePath,
+                text = it.readText(),
+                ruleSets = ruleSets,
+                userData = userData,
+                debug = debug,
+                editorConfigPath = additionalEditorConfig,
+                script = !it.name.endsWith(".kt", ignoreCase = true),
+                cb = { lintError, isCorrected ->
+                    errors.add(SerializableLintError(lintError) to isCorrected)
+                }
+            )
+
+            if (formatSource) {
+                val currentFileContent = it.readText()
+                val updatedFileContent = KtLint.format(ktLintParameters)
+
+                if (updatedFileContent != currentFileContent) {
+                    it.writeText(updatedFileContent)
+                }
+            } else {
+                KtLint.lint(ktLintParameters)
+            }
+            result.add(
+                LintErrorResult(
+                    lintedFile = it,
+                    lintErrors = errors
+                )
+            )
+        }
+
+        ObjectOutputStream(
+            BufferedOutputStream(
+                FileOutputStream(
+                    parameters.discoveredErrorsFile.asFile.get()
+                )
+            )
+        ).use {
+            it.writeObject(result)
+        }
+    }
+
+    private fun generateUserData(): Map<String, String> {
+        val userData = mutableMapOf(
+            "android" to parameters.android.get().toString()
+        )
+        val disabledRules = parameters.disabledRules.get()
+        if (disabledRules.isNotEmpty()) {
+            userData["disabled_rules"] = disabledRules.joinToString(separator = ",")
+        }
+
+        return userData.toMap()
+    }
+
+    private fun loadRuleSetsAndFilterThem(
+        enableExperimental: Boolean,
+        disabledRules: Set<String>
+    ): Set<RuleSet> = loadRuleSetsFromClasspath()
+        .filterKeys { enableExperimental || it != "experimental" }
+        .filterKeys { !(disabledRules.contains("standard") && it == "\u0000standard") }
```

Comment:
Why would we get a string starting with the null character here?

Reply:
It is copy-paste from KtLint CLI: https://github.com/pinterest/ktlint/blob/aa03a6b6b74cf71d8c2ede0c0a5e2305ce2e4327/ktlint/src/main/kotlin/com/pinterest/ktlint/internal/RuleSetsLoader.kt#L24

Initially it was introduced by Shyiko to enforce "standard" ruleset to be always the first.